gapic: add default method deadlines

noahdietz opened the pull request:
This adds consumption of the `timeout` in the generation config, using it to impose a default `context` deadline on client methods. This is an _overall deadline_ that encompasses all _RPC attempts_ made by the method.

Rough design:
* A `context` deadline set by a user is respected always
* the context derived from adding a deadline is canceled in the following scenarios:
  * in the event of an error in any method on the top-level client ~or a stream-wrapper client~
  * when a Unary method returns (using `defer`)
  * ~when a server-streaming client wrapper `Recv` receives an `io.EOF`~
  * ~when a client-streaming client wrapper `CloseAndRecv`~
  * ~when a bi-directional streaming client wrapper has both ends (client send & recv) of the client closed (via `CloseSend` & `Recv`-with-`io.EOF` respectively)~
* ~streaming methods return a generated wrapper that handles cancelation logic and passes through to the gRPC stub for all other work~

_Note: this will **not** support methods that return an `Iterator` at the outset. There needs to be more discussion around Iterator vs. per-Iteration deadlines._ 

_Update: this will **not** support streaming methods at the outset. There needs to be more discussion around the use case. The corresponding code was removed from this PR._

_Update: there will be an opt-out environment variable that's generated too._

Addresses part of #411

codyoss commented on internal/gengapic/testdata/method_BidiThings.want:
```diff
@@ -1,15 +1,71 @@
 	}, opts...)
 	if err != nil {
+		if cancel != nil {
+			cancel()
```
Are there cases where the context will have not been canceled by this point already?

noahdietz replied:
> Should this also be canceled for the happy path?

This same question came up in the doc. This specific `cancel` call is for when `gax` returns an error while executing `Invoke` i.e. during stream creation. The happy path is covered by the follwing stream client wrapper (e.g. `fooBidiThingsClient`) that `cancel`s in the success scenarios described in the description (and in the doc).

Does that help?

tbpg commented on internal/gengapic/gengapic.go:
```diff
@@ -481,6 +489,34 @@ func (g *generator) emptyUnaryCall(servName string, m *descriptor.MethodDescript
+		return
+	}
+
+	g.printf("if _, set := ctx.Deadline(); !set {")
```
Nit: I think `ok` is usually used as the variable name? Probably doesn't matter.

noahdietz replied:
I'll go with `ok` to stick to common practice. Done.

tbpg commented on internal/gengapic/gengapic.go:
```diff
@@ -481,6 +489,34 @@ func (g *generator) emptyUnaryCall(servName string, m *descriptor.MethodDescript
+	}
+
+	g.printf("if _, set := ctx.Deadline(); !set {")
+	g.printf("  cc, cancel := context.WithTimeout(ctx, %d * time.Millisecond)", t)
```
Nit: Perhaps `cctx`?

noahdietz replied:
Yeah, that's better! Done.

noahdietz commented:
> Can we do any sort of extra (?) leak detection here? 

Yeah, I agree, we should. I will take a look at adding leak detection to the Showcase-based integration tests. I've never done that before, but I see there might be an example in google-cloud-go "leackcheck"?

noahdietz commented:
Opened #426 to add basic `leakcheck` to the integration tests. I will rebase this change once that is merged in order to help verify.

noahdietz commented:
I rebased this PR to include the addition of the `leakcheck` tool in our integration test module.

tbpg commented:
Thanks! LGTM!

noahdietz commented:
Rebased the branch to pull in the retry/deadline test suite from #428 and remove the default deadline changes for streaming methods. 

Added the opt-out env var code gen and a test.

Added more default deadline integration tests.

tbpg commented on internal/gengapic/gengapic.go:
```diff
@@ -47,6 +47,7 @@ const (
 	paramError = "need parameter in format: go-gapic-package=client/import/path;packageName"
 	alpha      = "alpha"
 	beta       = "beta"
+	disableDeadlinesVar = "GOOGLE_API_GO_EXPERIMENTAL_DISABLE_DEFAULT_DEADLINE"
```
Is this a pattern for env vars for this repo? I'm OK with it, just curious.

noahdietz replied:
Pattern for the env var name? or for how I'm implementing it? The name was suggested by cbro.

noahdietz replied:
To be clear, this is the name of the env var that _generated clients_ will load for the opt-out

tbpg replied:
Yeah, I wasn't sure if `GOOGLE_API_GO_EXPERIMENTAL_` was a known, common, prefix for env vars for clients. Just looked and I don't see it anywhere. So, I don't think so.

noahdietz replied:
Yeah, I looked around google-api-go-client and google-cloud-go for any sort of pattern in env var naming that we could follow here...there was none, all kind of adhoc.
